### PR TITLE
fix(datastore): preserve ssr context

### PR DESCRIPTION
**Pull request by iartemiev**
#### Description of changes

The Amplify library exports the `withSSRContext` function (https://docs.amplify.aws/lib/ssr/q/platform/js/#withssrcontext) that allows a Next.js application to use client-side credentials in an isolated context on the server. The function creates new instances of the Auth, API, Credentials, and DataStore classes and configures this instance of Auth with the client-side cookies injected into it from the client-side request.

Currently, a library bug is preventing DataStore from using these instances of Auth and API while running in an SSR Context. The reference to the instance does not get preserved in the DataStore implementation. Therefore, DataStore simply imports new default instances of these classes lacking the client-side auth configuration. This breaks all SSR flows that depend on client-side credentials, such as the ones in the linked issues.

This PR adds a context object that stores references to Auth and API categories in the top-level DataStore class (`datastore.ts`) and passes the context object by reference to the descendent classes that use these categories. This allows DataStore to perform authenticated requests while running in the SSR context, fixing the linked GH issues.

#### Issue #, if available

https://github.com/aws-amplify/amplify-js/issues/8792
https://github.com/aws-amplify/amplify-js/issues/10027

#### Description of how you validated changes
1. Created new sample app and test script that validates DataStore with owner Auth with SSR Context: https://github.com/aws-amplify/amplify-js-samples-staging/pull/366 - **_must be merged before merging this PR_**
2. Ran all integ tests (including the new one, integ_next_datastore_owner_auth) against the changes in this branch: https://app.circleci.com/pipelines/github/aws-amplify/amplify-js/12228/workflows/5e01d94a-0456-4448-bdff-f7bdb320b5da
3. Updated unit test for `ssrContext`
4. Tested locally with additional sample apps

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by dpilch — APPROVED** [review]

File: packages/datastore/src/authModeStrategies/multiAuthStrategy.ts [hunk 1]
```diff
@@ -121,29 +122,31 @@ function getAuthRules({
...
-		currentUser = await Auth.currentAuthenticatedUser();
-	} catch (e) {
-		// No current user
-	}
+export const multiAuthStrategy: (
+	amplifyContext: AmplifyContext
+) => AuthModeStrategy =
+	(amplifyContext: AmplifyContext) =>
+	async ({ schema, modelName }) => {
+		amplifyContext.Auth = amplifyContext.Auth || Auth;
```
Comment (elorzafe) [inline packages/datastore/src/authModeStrategies/multiAuthStrategy.ts:130]: You dont need L1 anymore?

File: packages/datastore/src/datastore/datastore.ts [hunk 2]
```diff
@@ -1417,6 +1433,10 @@ class DataStore {
 	};
 
 	configure = (config: DataStoreConfig = {}) => {
+		this.amplifyContext.Auth = this.Auth;
+		this.amplifyContext.API = this.API;
+		this.amplifyContext.Cache = this.Cache;
+
```
Comment (elorzafe) [inline packages/datastore/src/datastore/datastore.ts:1439]: Is this something that can have implications on how you are importing the library?

For example the order of the imports or if you are when is `amplifyContext` been modified?

**Review by elorzafe — COMMENTED** [review]
Just a general question related about `amplifyContext`, I think that gets generated when doing `withSSRContext`, but what happens if you dont do that? or if someone calls that after doing `DataStore.start`?

**Comment by iartemiev** [comment]
> Just a general question related about `amplifyContext`, I think that gets generated when doing `withSSRContext`, but what happens if you dont do that? or if someone calls that after doing `DataStore.start`?

@elorzafe with the change in this PR, `amplifyContext` always gets created. But it's a no-op if you're outside of the SSR Context. It'll just reference the default instances of Auth and API.

File: packages/datastore/src/authModeStrategies/multiAuthStrategy.ts (hunk 1 above)
Comment (iartemiev) [reply packages/datastore/src/authModeStrategies/multiAuthStrategy.ts:130]: What's L1 referring to here?

File: packages/datastore/src/datastore/datastore.ts (hunk 2 above)
Comment (iartemiev) [reply packages/datastore/src/datastore/datastore.ts:1439]: > Is this something that can have implications on how you are importing the library? For example the order of the imports 

No, I don't think so. It just retains the reference either to the default instance of the module or to the class instantiated inside of withSSRContext implementation. It doesn't matter how you import the library. 

> or if you are when is amplifyContext been modified?
`amplifyContext` doesn't get modified anywhere outside of the DataStore class. We pass it into the descendent classes as a `readonly` property.

**Review by svidgen — APPROVED** [review]
I think this is reasonable. Thanks for fixing this!

**Pull request merged**